### ISSUE TITLE
Wrong HT value when you use a coupon

### ISSUE THREAD

**Issue by InformatiqueProg** [thelia/thelia#1476]
![capture](https://cloud.githubusercontent.com/assets/7933327/7961716/778e8d22-0a09-11e5-90a7-44db196ce196.PNG)
With a coupon off 30%, it would be
ht => 47.83€
Tva 20% => 9.57€
Ttc => 57.40€

**Comment by bibich** [comment]
@Alban-io could you check this in your work on coupon.

**Comment by roadster31** [comment]
@bibich @Alban-io any update about this problem ?

**Comment by InformatiqueProg** [comment]
I have not tested this recently, but i think i have seen pass an PR.
@roadster31 you have the problem on Thelia 2.3.3 ?

**Comment by roadster31** [comment]
Yes, and even on master branch. See the example below (the coupon i a 10% discount). The HT price seems wrong

![img-2017-05-11_12-31-51](https://cloud.githubusercontent.com/assets/2197734/25945164/38420b24-3646-11e7-9723-19665f6594ed.png)

**Comment by InformatiqueProg** [comment]
I agree (Thelia 2.3.3), see order below
Ht is wrong  76.30€ (it's seem to be total products HT-discount Ttc => 43.35*2-10.4)
, it would be 78.03€  (total products HT-discount Ht)
TVA is wrong 17.34€ (is TVA without discount), it would be 15.60€.
![capture](https://cloud.githubusercontent.com/assets/7933327/25948222/d584085e-3652-11e7-824d-b43de026648b.JPG)

**Comment by InformatiqueProg** [comment]
@gillesbourgeat This PR seem resolve the problem, do you remember why you closed it ?
https://github.com/thelia/thelia/pull/1520

**Comment by InformatiqueProg** [comment]
Je prepare une PR pour le panier, 
par contre pour les commandes passées c'est pas possible on n'a plus les "taxe rules" juste des montant de tva...

**Comment by InformatiqueProg** [comment]
To resolve this for orders, i think we could add a column "amount_discount" in `order_product_tax`.
In this we would save the "tax amount discounted" and we 'll use it to correct the TVA and price without tax.
Another idea ?

**Comment by roadster31** [comment]
The taxes are stored in the `order_product_tax` table when a new order is created. Maybe we can use this table to recalculate taxes for past orders ?

**Comment by roadster31** [comment]
As there is already a `postage_tax` column in the order table, we may add a `discount_tax` column to store the tax amount delta. This value will be used in Thelia\Model\Order::getTotalAmount() to get the right amounts.

**Comment by zorn-v** [comment]
If you think about adding columns, tables etc, just remember about meta_data table and models with convenient methods for get/put data to it.

**Comment by InformatiqueProg** [comment]
@roadster31 You have got a better idea than me
i can work on it but not before 1 or 2 week...

**Comment by roadster31** [comment]
@zorn-v meta_data tables are not a good way to store relational data, as it is not possible to use foreign keys to create relations between tables, and benefit from Propeml ORM mapping, and MySQL delete cascade.

@InformatiqueProg okay. I'll try to work on it :)

**Comment by anoziere** [comment]
This was fixed in 2.6 back in 2020 (commit 3df0568a1, refined later by #3242 and #3325) via a proper average tax-factor calculation for order discounts, and the same fixed logic is present in the Thelia 3 rewrite (Model/Order.php). Closing as no longer reproducible on either branch.
